teps/0089-nonfalsifiable-provenance-support.md

```diff
@@ -224,24 +291,95 @@ The entrypointer image should also be able to see that this flag is set and acco
+
+Each TaskRun status that is written by the TektonController will be signed to ensure that there is no external
+tampering of the TaskRun status. Upon each retrieval of the TaskRun, the TektonController checks if the status is initialized,
+that the signature in the annotations validate the current status.
 The signature and SVID will be stored as annotations on the TaskRun, and can be verified by a client.
 
 ### Verification in Chains
-Once Signed Results are available, we'll add verification of Signed Results to Chains.
+Once Signed TaksRuns are available, we'll add verification of Signed TaskRun to Chains.
 If verification fails for a TaskRun then Chains will not sign it.
 
-
 ## Test Plan
 Tests for Signed Results:
 1. Enabling the alpha feature for SPIRE in Tekton
-1. Requesting an SVID & signature over Results for a TaskRun
-1. Verification of SPIRE with Chains
-1. Verify that a TaskRun that isn't created by Tekton isn't signed Chains
+1. Verify that a necessary fields are present in the pod status result
+1. Verify that a TaskRun pod status being modified with incorrect results isn't verified by the TektonContoller
 
 Tests for Signed TaskRuns:
-1. Verify that a TaskRun that's been modified during execution isn't verified
-1. Verify that a TaskRun that's been modified after execution isn't verified
-
+1. Verify that a TaskRun that's been modified during execution isn't verified by the TektonController
+1. Verify that a TaskRun that's been modified after execution isn't verified by the TektonController
+1. Verify that a TaskRun that's been modified during execution isn't verified by the Chains
+1. Verify that a TaskRun that's been modified after execution isn't verified by the Chains
+
+## Threat Model
+
+There are 5 main components in this threat model:
+- TektonController (Signer/Verifier)
+- Chains (Verifier)
+- TaskRun Pod/Entrypointer (Signer)
+- SPIRE Server (Authority)
+- SPIRE Agent (Workload Attestor)
+
+There are several capabilities that we want to ensure are correct:
+- Verification of signed Results are signed by the correct TaskRun Pod/Entrypointer step
+  - Verified by TektonController through SPIRE Authority
+- Verification of signed TaskRuns are signed by the TektonController
+  - Verified by TektonController and Chains through SPIRE Authority
+
+Due to the implementation of these different components being similar in nature of how they interact with the architecture (k8s, SPIRE, etc.), the two cases can be evaluated against the same threats.
+
+Threats vectors around verification fall into several categories:
+- Ability to create a verifiable false signature
+- Able to influence a skip of verification step
+- Able to influence verification to verify against a false authority
+
+### Ability to create a verifiable false signature
+
+Potential threats:
+- Access to SPIRE upstream authority lets an attacker sign arbitrary values.
+- Access to pod execution environment allows minting false signatures for that TaskRun step
```

Comment:
is this also possible via the steps in a Task themselves? i.e. can I write a Task that has a step that accesses the private key and makes it available externally or uses it to sign something?

Reply:
okay so im trying to imagine the worst thing that someone could do, and what ive come up with is

a. run a task that has steps that copy the cert + private key somewhere else
b. construct a fake taskrun that looks like it has already executed and produced results - include that cert and private key in the status of the fake taskrun
c. pipelines controller reconciles this fake taskrun, sees it has already finished executing, it now needs to verify this Taskrun (before adding a signature as an annotation - the fake taskrun could contain a spoofed annotation but it would be incorrect so the controller would set the signed annotation to failed)
d. the pipelines controller will try to verify the results in this fake taskrun, and to do that it will request the trust bundle from spire that should have been used to sign the result
e. the trust bundle spire returns either won't match or won't exist (since the taskrun didnt execute) - not sure which but either way the controller will be able to tell this taskrun is not legit

^^ does this sound accurate? i.e. even if you copy the private key (and cert) and make them publicly available, nothing interesting can be done with them outside of the pod that SPIRE has associated them with?

> I elaborate on this in the other comment about why the cert is part of the output.

im still trying to understand if the cert is required - it seems to me like as long as anyone verifying the signed results asks SPIRE for the trust bundle it should expect to be used, im still not seeing how having access to the cert is required

> The check is done to make sure that the signature is for this particular step, so one is not able to sign on behalf of others.

probably a nit but: the cert/trust bundle would be for the entire taskrun, not for a particular step in the taskrun, is that right?

Reply:
Talked with @lumjjb a bit offline and I think I finally get it now: I hadn't actually understood what the contents of the cert were X'D PLUS the additional detail that we need to keep the public key (contained in the cert) around for verification b/c it is short lived

since the cert is signed by the Trust Bundle, we can make sure that we can trust the cert (and the public key it contains), and as long as the tekton controller can verify that the cert is appropriate for the taskrun to be using (i.e. the ID of the taskrun matches the ID inside the cert) we should be gtg!